### PR TITLE
PS-7746 - Manage deallocation of `ha_innobase::m_share` correctly

### PR DESCRIPTION
**Pull request by ldonoso**
https://ps57.cd.percona.com/view/5.7/job/percona-server-5.7-param/471/

---
When `free_share(m_share)` is called, we need to:

- Assign it the NULL value.
- Call `dict_table_close`

### CODE REVIEW

File: storage/innobase/handler/ha_innodb.cc [hunk 1]
```diff
@@ -7030,8 +7041,8 @@ ha_innobase::open(
 		if (UNIV_UNLIKELY(m_share->ib_table &&
 				  m_share->ib_table->is_corrupt &&
 				  srv_pass_corrupt_table <= 1)) {
-			free_share(m_share);
-
+			free_share_and_nullify(&m_share);
+			dict_table_close(ib_table, FALSE, FALSE);
```
Comment (venkatesh-prasad-v) [inline storage/innobase/handler/ha_innodb.cc:7045]: I think this should be
```suggestion
			dict_table_close(m_share->ib_table, FALSE, FALSE);
```

Comment (satya-bodapati) [reply storage/innobase/handler/ha_innodb.cc:7045]: Consider the intrinsic table opening, I would say ib_table is fine

File: storage/innobase/handler/ha_innodb.cc [hunk 2]
```diff
@@ -7066,6 +7066,7 @@ ha_innobase::open(
 			  srv_pass_corrupt_table <= 1)) {
 
 		free_share(m_share);
+		dict_table_close(ib_table, FALSE, FALSE);
```
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc:7079]: I think this is good idea to call dict_table_close() on error paths. We have increased the ref_count in open_table() and for intrinsic tables, we did a acquire() which increases ref_count.

dict_table_close() can handle intrinsic tables too. But we have to be careful in passing NULL to dict_table_close(). Some paths set ib_table to NULL. and following dict_table_close(ib_table) can cause crash.

Just analyze and ensure dic_table_close(NULL) doesn't happen.

Comment (ldonoso) [reply storage/innobase/handler/ha_innodb.cc:7079]: Yes, I have double checked that `dict_table_close` is provided a not null `ib_table`.

File: storage/innobase/handler/ha_innodb.cc [hunk 3]
```diff
@@ -1319,6 +1319,17 @@ free_share(
 /*=======*/
 	INNOBASE_SHARE*	share);		/*!< in/own: share to free */
 
+
+/************************************************************************/ /**
+Calls free_share and assign NULL to share. */
```
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]: it is not necessary to use old decoration
you could write the new doxygen comment header in 5.7 too

```
/** Calls free_share and assign NULL to share
@param[in,out]    share  table share to free */
```

File: storage/innobase/handler/ha_innodb.cc (hunk 2 above)
Comment (satya-bodapati) [reply storage/innobase/handler/ha_innodb.cc:7079]: Great!

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/handler/ha_innodb.cc [hunk 4]
```diff
@@ -1320,8 +1320,8 @@ free_share(
 	INNOBASE_SHARE*	share);		/*!< in/own: share to free */
 
 
-/************************************************************************/ /**
-Calls free_share and assign NULL to share. */
+/* Calls free_share and assign NULL to share
```
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]: /**  Calls
(two stars)

**Review by satya-bodapati — APPROVED** [review]

**Comment by satya-bodapati** [comment]
@ldonoso Please update PR with squashed commit.

**Review by venkatesh-prasad-v — APPROVED** [review]
Approved! Please don't forget to squash the commits before merging.

**Review by satya-bodapati — COMMENTED** [review]
Please squash the commits luis for final approval

**Pull request merged**